spdx/parsers/tagvaluebuilders.py

```diff
@@ -955,6 +960,78 @@ def set_pkg_comment(self, doc, text):
         else:
             raise CardinalityError("Package::Comment")
 
+    def set_pkg_primary_package_purpose(self, doc, purpose):
+        """
+        Set the package's primary purpose.
+        Raise CardinalityError if more than one purpose is set.
+        Raise SPDXValueError if purpose is unknown.
+        """
+        self.assert_package_exists()
+        if not self.package_primary_purpose_set:
+            self.package_primary_purpose_set = True
+            for purpose_enum in PackagePurpose:
+                if purpose == purpose_enum.name:
+                    doc.packages[-1].primary_package_purpose = purpose_enum
+                    return True
+            else:
+                raise SPDXValueError("Package::PrimaryPackagePurpose")
+        else:
+            raise CardinalityError("Package::PrimaryPackagePurpose")
```

Comment:
You can convert this to an early return and reduce indentation:
```
if self.package_primary_purpose_set:
    raise CardinalityError("Package::PrimaryPackagePurpose")
<remaining logic>
```
(in the other methods as well)

Reply:
that was a lot of work, but it looks much cleaner now! :)